content/md/posts/2023-07-08-unit-testing.md

```diff
@@ -0,0 +1,66 @@
+
+This test allows you to prove that user registration for the website works. Furthermore, testing individual components allows problems to be diagnosed more quickly when debugging. Knowing which components work allows the faulty components to be identified and fixed.
+
+## How do I use unit tests in tandem with accepttance testing?
+
+With acceptance testing, you test that a collection of events occur in a specific result. If you want to test every result for all event combinations, you will end up with countless tests.
+
+Using a combination of unit testing and acceptance testing, you can still test certain event combinations while assuring that other event combinations would not occur.
+
+For example, If I had a website where the user sees their dashboard after logging in at the home page, I could have one acceptance test check that scenario. Given a user who heads to the home page and logs in with the correct credentials, they will then see their dashboard. To be confident that users who enter incorrect credentials don't see their dashboard, I would like to have some tests that can confirm this. However, repeating the earlier acceptance test with countless variations of incorrect credentials would be a lot of effort.
+
+On the other hand, If I know that the user only sees their dashboard if they are authenticated, I could instead test my function which validates the user with a slew of incorrect credentials. Since viewing the dashboard depends on the result of the validation function, proving the function does not validate the user given incorrect credentials shows that an unauthenticated user cannot view their dashboard.
+
+
+## A real life example
+
+I will show an example of a unit test from a document sharing site I am working on. In this test, I check whether a registered user can login with hardcoded credentials and whether their account information is updated accordingly.
+
+Here is the code for the test.
+
+```clojure
+(t/deftest registered-user-login
+  (rf-test/run-test-async
+   (routes/init-routes!)
+   (core/firebase-init!)
+   (rf/dispatch-sync [::events/initialize-db])
+   (rf/dispatch [::events/login])
+   (rf-test/wait-for [::events/fetch-documents] 
+    (let [user (rf/subscribe [::subs/user])
+          name (rf/subscribe [::subs/name])]
+      (t/are [x y] (= x y)
+        :registered @user
+        "han@skywalker.com" @name)))))
+```
+
+First, I initialize my routes, setup the database, and connect to firebase. After that, I login and wait for the user's documents to be fetched. I then check whether the name of the user matches the credentials entered earlier and if the user is noted as registered.
```

Comment:
Maybe a diagram here could work better... having said that... if the goal is to illustrate a unit test... maybe let's try to find a different example that doesn't include re-frame.